### PR TITLE
Get subject from cache

### PR DESCRIPTION
**Pull request by annyhe**
- this feature is behind a short term toggle
- extract filter, limit, and offset from cache/models/sample
- add tests for when cache is on: GET subjects filter, limit, and offset
- does NOT include the changing perspectives to get subjects from cache

### CODE REVIEW

File: cache/models/samples.js [hunk 1]
```diff
@@ -1010,7 +801,7 @@ module.exports = {
    * @returns {Promise} - Resolves to a list of all samples objects
    */
   findSamples(req, res, logObject) {
-    const opts = getOptionsFromReq(req.swagger.params);
+    const opts = modelUtils.getOptionsFromReq(req.swagger.params, helper);
```
Comment (annyhe) [inline cache/models/samples.js]: need the helper as a param, as helper can be subject or sample or another resource's helper.

File: cache/models/samples.js [hunk 2]
```diff
@@ -894,7 +685,7 @@ module.exports = {
       return checkWritePermission(aspectObj, currSampObj, userName);
     })
     .then(() => {
-      cleanQueryBodyObj(reqBody);
+      modelUtils.cleanQueryBodyObj(reqBody, sampleFieldsArr);
       let value = '';
```
Comment (annyhe) [inline cache/models/samples.js]: need sampleFieldsArr as a param, since cleanQueryBodyObj uses sampleFieldsArr

**Comment by annyhe** [comment]
FYI I tested this branch on master. Both master and this feature branch GET /subjects returns sorted by absolutePath.

**Comment by shriramshankar** [comment]
I will review this.

**Comment by annyhe** [comment]
with toggle turned off 
screen shot 2017-07-14 at 12 07 43 pm
with toggle turned on
screen shot 2017-07-14 at 12 15 23 pm

File: tests/cache/models/subjects/getAll.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+      expect(res.body.length).to.be.equal(2);
+      expect(res.body[0].absolutePath).to.be.equal(subject2);
+      expect(res.body[1].absolutePath).to.be.equal(subject1);
+      expect(Object.keys(res.body[1])).to.contain('name', 'absolutePath');
+      done();
+    });
+  });
+
+  it('trailing asterisk is treated as "starts with"', (done) => {
+    api.get(path + '?name=' + tu.namePrefix + '*')
```
Comment (pallavi2209) [inline tests/cache/models/subjects/getAll.js]: Can you please apply a filter which returns a subset and not all of them. Like? absolutePath = .....*

File: tests/cache/models/subjects/getAll.js [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+        done(err);
+      }
+
+      expect(res.body.length).to.be.equal(1);
+      expect(res.body[0].name).to.be.equal(subject1);
+      done();
+    });
+  });
+
+  it('offset filter', (done) => {
```
Comment (pallavi2209) [inline tests/cache/models/subjects/getAll.js]: I think the api links changes when we apply offset filter, Can you please check?

Comment (annyhe) [inline tests/cache/models/subjects/getAll.js]: they are CRUD endpoints, and do not change per offset.

File: tests/cache/models/utils.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      descArr = [
+        { name: '___Subject1', absolutePath: '___Subject1' },
+        { name: '___Subject2', 'absolutePath': '___Subject1.___Subject2' },
+        { name: '___Subject3', absolutePath: '___Subject1.___Subject3' },
+      ];
+    });
+
+    it('asc name', () => {
+      const result = utils.sortByOrder(descArr, ['name']);
+      expect(result).to.deep.equal(descArr.reverse());
```
Comment (annyhe) [inline tests/cache/models/utils.js]: revised descArr.

File: tests/cache/models/subjects/get.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    api.get(path + '/abc')
+    .set('Authorization', token)
+    .expect(constants.httpStatus.NOT_FOUND)
+    .end((err, res) => {
+      expect(res.body.errors[0].type).to.equal('ResourceNotFoundError');
+      done();
+    });
+  });
+
+  it('get by name, with fields filter', (done) => {
```
Comment (annyhe) [inline tests/cache/models/subjects/get.js]: done.

File: tests/cache/models/subjects/get.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
+      }
+
+      expect(res.body.name).to.be.equal(name);
+      expect(res.body.absolutePath).to.equal(name);
+      expect(Object.keys(res.body)).to.contain('name', 'absolutePath', 'id', 'apiLinks');
```
Comment (annyhe) [inline tests/cache/models/subjects/get.js]: done.

File: tests/cache/models/subjects/getAll.js [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+
+      expect(res.body.length).to.be.equal(3);
+      expect(res.body[0].absolutePath).to.be.equal(subject1);
+      expect(res.body[1].absolutePath).to.be.equal(subject2);
+      expect(res.body[2].absolutePath).to.be.equal(subject3);
+      done();
+    });
+  });
+
+  it('sort option desc', (done) => {
```
Comment (annyhe) [inline tests/cache/models/subjects/getAll.js]: done, with name

File: tests/cache/models/subjects/getAll.js (hunk 3 above)
Comment (annyhe) [inline tests/cache/models/subjects/getAll.js]: done

File: tests/cache/models/utils.js [hunk 9]
```diff
@@ -55,9 +55,9 @@ describe('cache utils test', () => {
     // sort is in-place. Hence need reset array for test independence.
     beforeEach(() => {
       descArr = [
-        { name: '___Subject1', absolutePath: '___Subject1' },
-        { name: '___Subject2', 'absolutePath': '___Subject1.___Subject2' },
         { name: '___Subject3', absolutePath: '___Subject1.___Subject3' },
+        { name: '___Subject2', 'absolutePath': '___Subject1.___Subject2' },
+        { name: '___Subject1', absolutePath: '___Subject1' },
```
Comment (pallavi2209) [inline tests/cache/models/utils.js]: I dont think the tests .to.deep.equal is checking the right thing. If I change the order of records in descArr, the tests will still pass. Can you please check whats wrong.

File: tests/cache/models/utils.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/cache/models/utils.js
+ */
+'use strict'; // eslint-disable-line strict
+
+const supertest = require('supertest');
```
Comment (pallavi2209) [inline tests/cache/models/utils.js]: supertest not required here

Comment (annyhe) [inline tests/cache/models/utils.js]: good catch. Removed

File: tests/cache/models/utils.js (hunk 9 above)
Comment (annyhe) [inline tests/cache/models/utils.js]: Yeah I replaced the deep equal with checking each entry in the result.

File: api/v1/controllers/subjects.js [hunk 11]
```diff
@@ -215,7 +216,25 @@ module.exports = {
...
+        }
+
+        u.logAPI(req, resultObj, response); // audit log
+        res.status(httpStatus.OK).json(response);
+      })
+      .catch((err) => u.handleError(next, err, helper.modelName));
+    } else {
+      doFind(req, res, next, helper);
+    }
   },
```
Comment (shriramshankar) [inline api/v1/controllers/subjects.js]: Looks like you missed changes to the getSubject method.

Comment (annyhe) [inline api/v1/controllers/subjects.js]: which changes?

Comment (annyhe) [inline api/v1/controllers/subjects.js]: Also, the populateSubjects gets only subjects that are published. So doFind may return different subjects. I talked to @iamigo about it and created another story to turn populateSubjects to return all subjects.

Comment (shriramshankar) [inline api/v1/controllers/subjects.js]: I am talking about doGet that is inside controllers/subjects.js

Comment (annyhe) [inline api/v1/controllers/subjects.js]: ok added getSubject in cache.

File: cache/models/subject.js [hunk 12]
```diff
@@ -186,4 +201,114 @@ function completeSubjectHierarchy(res, params) {
...
+        });
+      }
+
+      // add parentAbsolutePath
+      if (subject.parentAbsolutePath === undefined) {
+        subject.parentAbsolutePath = '';
+      }
+
+      // convert the strings into numbers
+      subject.childCount = parseInt(subject.childCount, 10) || 0;
```
Comment (annyhe) [inline cache/models/subject.js]: On master :
GET ALL subjects : the numeric fields: childCount, hierarchyLevel, are stringified and the parentAbsolutePath of root subject is undefined. https://refocus-devbox.herokuapp.com/v1/subjects?name=Salesforce

For GET subject: the same numeric fields are numbers, and for root subjects parentAbsolutePath=''. https://refocus-devbox.herokuapp.com/v1/subjects/Salesforce

This PR reproduces these inconsistencies. We can discuss what is the best course of action after I get back.

File: api/v1/controllers/subjects.js (hunk 11 above)
Comment (annyhe) [inline api/v1/controllers/subjects.js]: There is a clashing of opinions on where the cache/non-cache function splits off. @iamigo prefers to keep the cache-specific versions in the controllers file, whereas @shriramshankar prefers to push the cache vs non-cache implementation block into the helpers/verbs folder. This is not good as we are left with inconsistent code: ie. even in the same controller file, findSamples defines what to do when cache is turned on or off, whereas getSample calls doGet which then defines what to do when cache is turned on or off. It is better to decide on where to put the conditional and stick with it.

Comment (shriramshankar) [inline api/v1/controllers/subjects.js]: Yeah, we can decide on something and stick to it.

**Comment by annyhe** [comment]
@shriramshankar ignore what I mentioned about the differences between GET ALL and GET subject results on master. They are identical: childCount and hierarchyLevel are integers, while parentAbsolutePath is '' for the root subject.

Here is the PR deployed on devbox with the toggle turned on: https://refocus-devbox.herokuapp.com/v1/subjects/Airports
and
https://refocus-devbox.herokuapp.com/v1/subjects?name=Airports

**Comment by shriramshankar** [comment]
Changes look good. Will merge it later tomorrow.

**Pull request merged**